### PR TITLE
(RK-290) Add deprecation warning about duplicate module names

### PR DESCRIPTION
**Pull request by ajroetker**
This commit backports the duplicate module names error message code to
2.5.x as a deprecation warning.

### CODE REVIEW

File: lib/r10k/puppetfile.rb [hunk 1]
```diff
@@ -69,6 +69,17 @@ def set_forge(forge)
     @forge = forge
   end
 
+  # @param [Array<String>] modules
+  def validate_no_duplicate_names(modules)
```
Comment (andersonmills) [inline lib/r10k/puppetfile.rb:74]: I think you forgot to actually call this on load!

**Review by andersonmills — CHANGES_REQUESTED** [review]

File: lib/r10k/puppetfile.rb (hunk 1 above)
Comment (ajroetker) [reply lib/r10k/puppetfile.rb:74]: Fixed up!

**Review by andersonmills — APPROVED** [review]
```
roquefort:~/puppet/src/r10k (PR724)
$ be r10k puppetfile check
WARN     -> Puppetfiles should contain duplicate module names and will result in an error in r10k v3.x. Please remove the remove the duplicates of the following modules:
motd
Syntax OK
```

File: lib/r10k/puppetfile.rb [hunk 2]
```diff
@@ -69,6 +70,16 @@ def set_forge(forge)
...
+    dupes = modules
+            .group_by { |mod| mod.name }
+            .select { |_, v| v.size > 1 }.map(&:first)
+    unless dupes.empty?
+      logger.warn _("Puppetfiles should contain duplicate module names and will result in an error in r10k v3.x. Please remove the remove the duplicates of the following modules:\n%{dupes}" % {dupes: dupes.join("\n")})
```
Comment (andersonmills) [inline lib/r10k/puppetfile.rb]: I now think these \n's should be spaces, and this should all be on one line in the log.

**Review by andersonmills — CHANGES_REQUESTED** [review]

File: lib/r10k/puppetfile.rb (hunk 2 above)
Comment (haus) [reply lib/r10k/puppetfile.rb]: "Please remove the remove the " => "Remove the"

Comment (haus) [reply lib/r10k/puppetfile.rb]: the log message itself should be split into two `_()` calls though.

File: lib/r10k/puppetfile.rb [hunk 3]
```diff
@@ -69,6 +70,17 @@ def set_forge(forge)
...
   end
 
+  # @param [Array<String>] modules
+  def validate_no_duplicate_names(modules)
+    dupes = modules
+            .group_by { |mod| mod.name }
+            .select { |_, v| v.size > 1 }.map(&:first)
+    unless dupes.empty?
+      logger.warn _('Puppetfiles should contain duplicate module names and will result in an error in r10k v3.x.')
+      logger.warn _("Please remove the remove the duplicates of the following modules: %{dupes}" % {dupes: dupes.join(" ")})
```
Comment (haus) [inline lib/r10k/puppetfile.rb]: "Please remove the remove the " => "Remove the "

**Comment by ajroetker** [comment]
@andersonmills should be fixed up

**Comment by ajroetker** [comment]
corresponding change to the master branch @andersonmills https://github.com/puppetlabs/r10k/pull/727

**Review by haus — APPROVED** [review]
LGTM

**Review by andersonmills — APPROVED** [review]
Well, I *thought* it was good.

File: lib/r10k/puppetfile.rb [hunk 4]
```diff
@@ -69,6 +70,17 @@ def set_forge(forge)
     @forge = forge
   end
 
+  # @param [Array<String>] modules
+  def validate_no_duplicate_names(modules)
+    dupes = modules
+            .group_by { |mod| mod.name }
+            .select { |_, v| v.size > 1 }.map(&:first)
+    unless dupes.empty?
+      logger.warn _('Puppetfiles should contain duplicate module names and will result in an error in r10k v3.x.')
```
Comment (andersonmills) [inline lib/r10k/puppetfile.rb]: Hah! "Puppetfiles should contain duplicate module names"

You forgot a "not"

**Review by andersonmills — CHANGES_REQUESTED** [review]

**Review by andersonmills — APPROVED** [review]

**Pull request merged**